primitives/api/proc-macro/src/decl_runtime_apis.rs

```diff
@@ -452,28 +344,60 @@ fn generate_runtime_decls(decls: &[ItemTrait]) -> Result<TokenStream> {
 
 		let call_api_at_calls = generate_call_api_at_calls(&decl)?;
 
-		// Remove methods that have the `changed_in` attribute as they are not required for the
-		// runtime anymore.
+		let trait_api_version = get_api_version(&found_attributes).unwrap_or(1);
+
+		let mut methods_by_version: BTreeMap<u64, Vec<TraitItemMethod>> = BTreeMap::new();
+
+		// Process the items in the declaration. The filter_map function below does a lot of stuff
+		// because the method attributes are stripped at this point
 		decl.items = decl
 			.items
 			.iter_mut()
 			.filter_map(|i| match i {
 				TraitItem::Method(ref mut method) => {
-					if remove_supported_attributes(&mut method.attrs)
-						.contains_key(CHANGED_IN_ATTRIBUTE)
-					{
+					let method_attrs = remove_supported_attributes(&mut method.attrs);
+					let mut method_version = trait_api_version;
+					// validate the api version for the method (if any) and generate default
+					// implementation for versioned methods
+					if let Some(version_attribute) = method_attrs.get(API_VERSION_ATTRIBUTE) {
+						let method_api_ver =
+							parse_runtime_api_version(version_attribute).unwrap_or(1);
+						if method_api_ver < trait_api_version {
+							let span = method.span();
+							let method_ver = method_api_ver.to_string();
+							let trait_ver = trait_api_version.to_string();
+							result.push(quote_spanned! {
+								span => compile_error!(concat!("Method version `",
+								#method_ver,
+								"` is older than (or equal to) trait version `",
+								#trait_ver,
+								"`. Methods can't define versions older than the trait version."));
+							});
+						} else {
+							// save method version
+							method_version = method_api_ver;
+						}
+					}
+
+					// Remove methods that have the `changed_in` attribute as they are not required
+					// for the runtime anymore.
+					if method_attrs.contains_key(CHANGED_IN_ATTRIBUTE) {
 						None
 					} else {
 						// Make sure we replace all the wild card parameter names.
 						replace_wild_card_parameter_names(&mut method.sig);
+
+						// partition methods by api version
+						methods_by_version.entry(method_version).or_default().push(method.clone());
+
 						Some(TraitItem::Method(method.clone()))
```

Comment:
The default trait declaration should only contain all methods that are "stable" for this version? 

And the default trait could just be an alias to the actual versioned trait, because they will be the same.

So, if we are at version 2, we will have the following code:

```
pub type MySuperRuntimeApi = MySuperRuntimeApiVersion2;
```